### ISSUE TITLE
I/O module: bug with specifying a different start for the channel than 0

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#20449]
### Summary of Problem
I discovered that if a file already has contents and you create a writing channel (at least using `openwriter`, though I don't see why `file.writer` would behave any differently since `openwriter` calls it), specifying a starting location other than 0 will cause all the characters prior to your starting location to be re-interpreted as binary.

This seems incorrect - I would expect the contents of the file outside of the specified region to be untouched, enabling concurrent modifications to distinct parts of the file.

### Steps to Reproduce

**Source Code:**

```chapel
use IO;

var filename = "openwriterLimited.txt";
var writeCh = openwriter(filename);
writeCh.writeln("She said she was looking for me");
writeCh.close();

var writeChStart = openwriter(filename, start=31);
writeChStart.write("us");
writeChStart.close();
```

**Compile command:**
`chpl foo.chpl`

**Execution command:**
`./foo`

**Associated Future Test(s):**
[`test/library/standard/IO/openwriterStartBug.chpl`](https://github.com/chapel-lang/chapel/blob/main/test/library/standard/IO/openwriterStartBug.chpl) #20451 

### Configuration Information

- Output of `chpl --version`: chpl version 1.28.0 pre-release
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: clang
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: none *
CHPL_AUX_FILESYS: none
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
Apple clang version 11.0.3 (clang-1103.0.32.62)
Target: x86_64-apple-darwin20.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

**Comment by cassella** [comment]
This issue and #20450 look like the file has been truncated before the test writes data to it.  (There, truncating the file and writing three bytes; here truncating, seeking, and writing three bytes leaving the seeked-over region with NULL bytes.)

Following the `openwriter()` call to `openwriterHelper()`

```chapel
  var fl:file = try open(path, iomode.cw);
```

where the table in IO.chpl says `iomode.cw` is equivalent to the `fopen()` argument `"w"`.  And indeed `_modestring()` converts `iomode.cw` to `"w"`.  The `fopen(3)` man page description of `"w"` includes

```
Truncate file to zero length or create text file for writing.
```

I don't see an `fopen()` mode that will open the file for writing, creating it if it doesn't exist, but also not truncating it if it does exist.  `"r+"` will open it for reading and writing without truncating, but won't create the file if it doesn't exist.

`openwriter()`'s comment says "This function is equivalent to calling :proc:`open` with ``iomode.cwr`` ...", and `iomode.cwr`'s doc refers to `iomode.cw` which says it truncates the file.    So I think the truncation is consistent with the documentation.


(Aside: Since `openwriter()` ends up using `cw`, it's not quite equivalent to `cwr`.  The `cwr` open will fail if the process lacks read access to the file (e.g. it's write-only, say mode `0200`).  The `cw` open it uses will succeed in this case.)

(Aside 2: the strings `_modestring()` returns are defined before the function, with a comment
```
// these strings are here (vs in _modestring)
// in an attempt to avoid string copies, leaks,
// and unnecessary allocations.
```

Are those still concerns?  That comment was introduced in 2012's f386efb3)

**Comment by lydia-duncan** [comment]
That would explain why truncating the output sounded familiar to me.  However, I don't think that addresses the concern that this means multiple writing channels cannot be open to the same file in parallel, which was something it sounded like we wanted.  But perhaps that means we shouldn't expect to support such things?

**Comment by mppf** [comment]
@lydia-duncan - if you don't use `openwriter`, you should be able to open the file once, and then have multiple writing channels that refer to the same file.

It is a different feature to support opening the file multiple times & having one channel per open file. But it would be reasonable to expect that to work as well & AFAIK it could work if the truncating issue is resolved (by making a separate `open` call).

**Comment by lydia-duncan** [comment]
If openwriter is going to destroy the previous contents of the file and that doesn't seem like something we want, does it make sense to provide the region argument for it?

**Comment by lydia-duncan** [comment]
In the IO subteam meeting we had yesterday, we discussed two ways to avoid this problem:
1. We could add a `mode` arg to `openwriter`, to allow the user to specify what mode they want their file to be opened in.  It seems as though `iomode.rw` would be a better fit if the user wanted to keep at least some of the original contents of the file.  We could make that value the default, which might result in less surprises for users.
- Pros:
  - gives the user more control, perhaps they do actually want to overwrite the contents to that point in the file or truncate it more explicitly
  - makes it a little more similar to the pattern it is simplifying (`open(...).writer(...)`)

- Cons:
  - Don't really anticipate needing to do the same for `openreader`, its counterpart, which would make their interface differ somewhat
  - Extension of functionality when 2.0 is more aimed for stabilization.  Would need development and testing to ensure it works properly 

2. We could remove the arguments that adjust the position in the file (currently `start` and `end` but without this issue would have been adjusted to be `region` like the other functions listed in #20133).  This could make it more apparent that the entire file is being overwritten by the channel that gets opened.

- Pros:
  - Avoids the potential of a stumbling block for users
  - These arguments are much more heavily used in `file.writer`, our testing system had no instances of them being used with `openwriter`.  Likely minimal impact to user code
  - Already doesn't cover full functionality of `open(...).writer(...)`, it's a convenience and should only provide what is useful

- Cons:
  - `openreader` has no problem with these arguments, so no need to remove them there.  This would be an interface difference between them (although there are similarly no examples of calling `openreader` with those arguments in our testing system, so we could also remove it there.  That seems unnecessarily restrictive)
  - Makes it less similar to `open(...).writer(...)`, which it is a convenience for.

The IO subteam would like to propose following the second of these two strategies.

**Comment by vasslitvinov** [comment]
If I am invoking `openwriter` with non-trivial start and end arguments (as separate args or as a bounded range), my intent is to mess with only that region of the file. I surely expect the rest of the file to remain in tact.

This contradicts the definition of `openwriter`, however, whose [chpldoc](https://chapel-lang.org/docs/master/modules/standard/IO.html?highlight=openwriter#IO.openwriter) says that it opens the file with the mode `cwr`. Which I read as "delete any pre-existing contents".

So if I `openwriter` with the range 5..10, what do we expect for the first five bytes of the file?

**Comment by lydia-duncan** [comment]
Under proposal 1, I would expect the behavior to vary depending on which mode is sent - with `rw`, it would be left alone, with `cwr` or `cw`, I think we would expect the first 5 bytes to be zero'd and the file to end after whatever was written.  Under proposal 2, it won't be possible to open with a reduced region.

**Comment by jeremiah-corrado** [comment]
I like proposal 2 better. 

I think it's nice to have a an interface with a separate *easy-mode* and *expert-mode*. Removing the extra args from `openwriter` will make the documentation that much less daunting to parse, which I think makes it more usable/approachable as a convenience function.

**Comment by bradcray** [comment]
I'm good with option 2, though this whole issue definitely made me wonder again whether we should just have users use `open` + `writer` rather than `openwriter`.  That said, I also buy Jeremiah's "it's a simpler convenience function" argument.
